### ISSUE TITLE
Standardize how we iterate over objects (`for`/`of`)

### ISSUE THREAD

**Issue by markocikos** [liferay/liferay-frontend-projects#354]
tl;dr: See below for discussion, but [plan](https://github.com/liferay/liferay-frontend-projects/issues/354#issuecomment-762117930) is:

- Turn _on_ `no-for-of-loops/no-for-of-loops` rule on older branches (7.3.x and prior) which must support IE, via explicit setting in `.eslintrc.js`.
- Turn the rule _off_ in the default config provided by `@liferay/eslint-config` (note that we'll probably continue to _bundle_ the `no-for-of-loops` plug-in with the config in order to not break existing set-ups that may be relying on it transitively).
- Update `@liferay/eslint-config` everywhere, freeing the people and bringing peace to the galaxy (we don't need to do an immediate migration; we can write a guideline for this and let usage drift towards it organically — if expected progress doesn't materialize, we can make a lint).

---

I've seen all below used in portal, and there are probably others.

1. 
```
for (const key in obj) {
    if (hasOwnProperty(key)) {
        // do stuff
    }
}
```
2. 
```
for (const key in obj) {
    // do stuff, we don't care about fringe inheritance case
}
```
3.
```
Object.keys(obj).forEach(key => {
    // do stuff
});
```
4.
```
for (const key of Object.keys(obj)) {
    // do stuff
}
```
5.
```
for (const [key, value] of Object.entries(obj)) {
    // do stuff
}
```
6.
```
for (let i = 0, l = Object.keys(obj).length; i < l; i++) {
    // do stuff
}
```

Usage of `for..on` is currently flagged as warning in eslint, although (5) is my favorite iteration  method. What should be the recommended method?

**Comment by victorg1991** [comment]
I guess that now that we don't support ie11, we can remove the lint rule forbidding for..of loops, right?

**Comment by wincent** [comment]
Yes, @victorg1991, but we have to do it in a way that doesn't break things when the change gets backported to older branches (which still support IE). There are a few options for that, but the best one is probably turning the rule _on_ explicitly in an `.eslintrc.js` override on each of the older branches and _then_ turning the rule off in the default config provided by `@liferay/eslint-config`.

**Comment by victorg1991** [comment]
That sounds reasonable @wincent :D. Should I open a ticket somewhere or is this being tracked somewhere?

**Comment by wincent** [comment]
I think _this_ is the ticket; I'll update the title and description to reflect our "conclusion".

**Comment by kresimir-coko** [comment]
This is fixed, I think? @bryceosterhaus

**Comment by bryceosterhaus** [comment]
Yeah looks like this is safe to close.
